cpp/src/bitmask/bit_mask.cuh

```diff
@@ -14,143 +14,224 @@
   }
 
   /**
-   * @brief Set a specific element (host only)
+   *  @brief clear the specified bit in the bit mask in an unsafe manner.
+   *
+   *  This function clears the specified bit in an unsafe manner.  It assumes that
+   *  that the calling code guarantees a thread-safe context.  That is, either
+   *  the function is called from a block of serial code, or the data is distributed
+   *  among the threads such that no two threads could be updating a bit in the
+   *  same memory location concurrently.
    *
-   * @param[in] element_idx
+   *  @param[in,out]  valid         The bit mask to update
+   *  @param[in]      record_idx    The record index
    *
-   * @return the specified element
    */
-  __host__ gdf_error set_element_host(gdf_size_type element_idx, const bit_mask_t &element) {
-    return bit_mask::put_element(element, valid_ + element_idx);
+  template <typename T>
+  CUDA_HOST_DEVICE_CALLABLE
+  void clear_bit_unsafe(bit_mask_t *valid, T record_idx) {
+    static_assert(std::is_integral<T>::value, "Record index must be of an integral type");
+
+    return gdf::util::turn_bit_off(valid, record_idx);
   }
 
   /**
-   * @brief Getter for valid
+   *  @brief set the specified bit in the bit mask in an threadsafe manner
+   *
+   *  This function sets the specified bit in an threadsafe manner.  It uses
+   *  atomic memory operations to guarantee that no update interferes with
+   *  on another.
+   *
+   *  @param[in,out]  valid         The bit mask to update
+   *  @param[in]      record_idx    The record index
    *
-   * @return pointer to valid bit container
    */
-  __host__ __device__ bit_mask_t *get_valid() const {
-    return valid_;
+  template <typename T>
+  CUDA_DEVICE_CALLABLE
+  void set_bit_safe(bit_mask_t *valid, T record_idx) {
+    static_assert(std::is_integral<T>::value, "Record index must be of an integral type");
+
+    const gdf_size_type rec{gdf::util::detail::bit_container_index<bit_mask_t, T>(record_idx)};
+    const gdf_size_type bit{gdf::util::detail::intra_container_index<bit_mask_t, T>(record_idx)};
+
+    atomicOr( &valid[rec], (bit_mask_t{1} << bit));
   }
 
   /**
-   * @brief Get length
+   *  @brief clear the specified bit in the bit mask in an threadsafe manner
+   *
+   *  This function clear the specified bit in an threadsafe manner.  It uses
+   *  atomic memory operations to guarantee that no update interferes with
+   *  on another.
+   *
+   *  @param[in,out]  valid         The bit mask to update
+   *  @param[in]      record_idx    The record index
    *
-   * @return length of bit container in bits
    */
-  __host__ __device__ gdf_size_type length() const {
-    return bitlength_;
+  template <typename T>
+  CUDA_DEVICE_CALLABLE
+  void clear_bit_safe(bit_mask_t *valid, T record_idx) {
+    static_assert(std::is_integral<T>::value, "Record index must be of an integral type");
+
+    const gdf_size_type rec{gdf::util::detail::bit_container_index<bit_mask_t, T>(record_idx)};
```

Comment:
If something from the detail sub-namespace of a different namespace than the current one is needed, shouldn't it be moved out of detail?